node/pkg/watchers/evm/watcher.go

```diff
@@ -930,3 +959,220 @@ func (w *Watcher) SetWaitForConfirmations(waitForConfirmations bool) {
 func (w *Watcher) SetMaxWaitConfirmations(maxWaitConfirmations uint64) {
 	w.maxWaitConfirmations = maxWaitConfirmations
 }
+
+// ccqSendQueryResponse sends an error response back to the query handler.
+func (w *Watcher) ccqSendQueryResponse(logger *zap.Logger, req *query.PerChainQueryInternal, status query.QueryStatus, resp *query.EthCallQueryResponse) {
+	queryResponse := query.CreatePerChainQueryResponseInternal(req.RequestID, req.RequestIdx, req.Request.ChainId, status, resp)
+	select {
+	case w.queryResponseC <- queryResponse:
+		logger.Debug("published query response error to handler", zap.String("component", "ccqevm"))
+	default:
+		logger.Error("failed to published query response error to handler", zap.String("component", "ccqevm"))
+	}
+}
+
+func (w *Watcher) ccqHandleQuery(logger *zap.Logger, ctx context.Context, queryRequest *query.PerChainQueryInternal) {
+	ccqMaxBlockNumber := big.NewInt(0).SetUint64(math.MaxUint64)
+
+	// This can't happen unless there is a programming error - the caller
+	// is expected to send us only requests for our chainID.
+	if queryRequest.Request.ChainId != w.chainID {
+		panic("ccqevm: invalid chain ID")
+	}
+
+	switch req := queryRequest.Request.Query.(type) {
+	case *query.EthCallQueryRequest:
+		block := req.BlockId
+		logger.Info("received query request",
+			zap.String("block", block),
+			zap.Int("numRequests", len(req.CallData)),
+		)
+
+		timeout, cancel := context.WithTimeout(ctx, 5*time.Second)
+		// like https://github.com/ethereum/go-ethereum/blob/master/ethclient/ethclient.go#L610
+
+		var blockMethod string
+		var callBlockArg interface{}
+		// TODO: try making these error and see what happens
```

Comment:
Leaving a note to make an issue for this. We should move this to a helper function that returns some go struct for a "blockId" that has a type (hash or height) and then we can make this nicer and more testable. Or maybe at least an "isBlockHash" check that's more explicit (optional 0x, 64 hex chars)